README.Rmd

```diff
@@ -0,0 +1,191 @@
+---
+output: github_document
+author: "Mauro Lepore, `r Sys.time()`."
```

Comment:
should the author name be generalized for something like this?

Reply:
Yes, I'll remove my name. I intended to be a point of contact but the motivated reader can `git blame -- README.Rmd`.
No file chosen
Attach files by dragging & dropping, selecting or pasting them.
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
* __Single-inputs workflow__: When the number of portfolios and investors is
exactly one, we use a specific set of scripts optimized for this purpose. This
workflow is typically used to generate interactive reports. Because this
workflow is available [online](https://platform.transitionmonitor.com/start), it
has been referred to as "the online workflow", and the scripts as "the web tool
scripts". This is misleading because this workflow can run offline too.
* __Multiple-outputs workflow__: When the number of portfolios and/or investors
is not exactly one, we use a different set of optimized scripts. This workflow
is typically used to produce static (.pdf) reports. By contrast with the
so-called "online" workflow, this one has been usually referred to as "the
offline workflow"; again, this is misleading.
Comment on lines +37 to +47
 
@cjyetman cjyetman 9 hours ago Member
I think a discussion about this is going to need to be had beyond PR comments because this might be a long debate.

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
(CB).
2. Run Analysis: Merge portfolios with asset-level data and scenarios, then
group results at company, portfolio and regional level.
3. Present results: Present the results in a clear output format.
 
@cjyetman cjyetman 9 hours ago Member
Personally, I would like to detach this 3rd step from the process in the future. What I imagine is that running the PACTA analysis generates results (meaning data that has been merged and certain calculations done) based on the input portfolio and various options. After that, a user (or web server) may want to use those results to generate an interactive report, or maybe an Executive Summary PDF, or maybe just one interactive alignment plot, or maybe just do some manual analysis with it in R, or maybe some combination of those. But that's the point where users branch out into different things.

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
Note: Code chunks with prompts "`$`" and "`#>`" correspond to bash and R,
respectively.
Comment on lines +60 to +61
 
@cjyetman cjyetman 9 hours ago Member
I was confused by this at first because I was expecting $ and #> to precede the bash and R commands that you were showing, but in fact these only show preceding stdout feedback on the console.

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd

### Siblings

PACTA_analysis depends on other "siblings" of the PACTA family. They are all
 
@cjyetman cjyetman 9 hours ago Member
this is only really true if/when you want to do extra stuff with the results of PACTA_analysis... like create an interactive report

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd

PACTA_analysis depends on other "siblings" of the PACTA family. They are all
available at <https://github.com/2DegreesInvesting> and should live alongside
PACTA_analysis/, under the same parent directory. All siblings include an empty file called ".pacta", which you can use to find them programmatically.
 
@cjyetman cjyetman 9 hours ago Member
this is also something I would strongly prefer to get away from.... this is why I prefer to detach the what-do-I-do-with-the-results part of the process... then once you have the results from running PACTA_analysis, you could take that wherever you want and do with it whatever you want.... you could have the create_interactive_report somewhere totally different and run it on your results, wherever you decided to put them

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
siblings="$(dirname $(find .. -maxdepth 2 -empty -type f -name '.pacta'))"
for i in $siblings; do
  cd $i && echo $i
  git log --decorate --oneline -n 1
done
Comment on lines +80 to +84
 
@cjyetman cjyetman 9 hours ago Member
I actually don't know how to interpret the result of this to see that the repos are up to date, and therefore I suspect others will not as well

$ ../pacta-data
$ 13215f8 (HEAD -> master, tag: 0.0.14, origin/master) adding new peer files for Austria (#35)
$ ../create_interactive_report
$ aefc204 (HEAD -> master, tag: 0.0.14, origin/master) Merge pull request #380 from MonikaFu/375-format-scenario-names
$ ../StressTestingModelDev
$ 45f124b (HEAD -> master, tag: 0.0.14, origin/master) 211 litigation pf level (#220)
$ ../PACTA_analysis
$ 61af20b (HEAD -> generalize-workflow) Update date
@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd

## Computing environment

The required computing environment is available in [this docker
 
@cjyetman cjyetman 9 hours ago Member
The docker image is not truly a "required" computing environment, and that contradicts the suggestion above that the web tools can and may be used in different context than what it's primarily intended for.

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
The required computing environment is available in [this docker
image](https://github.com/2DegreesInvesting/docker/tree/master/r-packages).

Other computing environments may be defined as workflows for GitHub actions.
 
@cjyetman cjyetman 9 hours ago Member
I had to read this a few times. I think it's not super clear what you're trying to convey here.

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
```{bash comment="$"}
ls .github/workflows
```
Comment on lines +101 to +103
 
@cjyetman cjyetman 9 hours ago Member
I don't understand what I'm meant to take away from the results of this command

ls .github/workflows
$ R-CMD-check.yaml
$ pr-commands.yaml
$ source-web-tool-scripts.yaml
@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
ls .github/workflows
```

Unless otherwise indicated, R packages should be up to date.
 
@cjyetman cjyetman 9 hours ago Member
How/where might that be indicated? By who/what? For what reason?

Also, is this a suggestion that the user should manually check that their R packages are up to date, or just a statement that they "should" be so the user doesn't have to worry about it?

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
<details> <summary>Session information.</summary>
```{r}
devtools::session_info()
```
</details>
Comment on lines +107 to +111
 
@cjyetman cjyetman 9 hours ago Member
not sure what this is for

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
* I move to the parent directory of PACTA_analysis and its siblings.

```bash
cd ~/git
 
@cjyetman cjyetman 9 hours ago Member
this is locally/user specific... maybe need to make that clear

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
cd ~/git
```

* I run (`docker run`) an ephemeral (`--rm`), interactive (`-ti`) container of
 
@cjyetman cjyetman 9 hours ago Member
I would take a wild guess and say that probably less than 50% of the people at 2dii have docker installed on their system... so maybe that needs to be listed up from as a requirement?

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
directory (`-w`) to PACTA_analysis/.

```{r comment="", echo=FALSE}
url <- "https://raw.githubusercontent.com/maurolepore/bin/master/docker-run-2dii-r-packages"
 
@cjyetman cjyetman 9 hours ago Member
I would rather see this code spelled out. Loading a script from a web URL and blindly running it sends chills down my spine

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
lines <- readLines(tmp)
lines <- lines[2:length(lines)]
writeLines(lines)
 
@cjyetman cjyetman 9 hours ago Member
that seems like a lot of hoop jumping just to display the command? I don't really understand what the purpose of that is

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
* Then I use R to produce the very file you are now reading.

```r
Rscript -e "rmarkdown::render('README.Rmd', output_format = 'md_document')"
 
@cjyetman cjyetman 9 hours ago Member
technically, Rscript is a shell command, not R

I know from experience, but I think maybe it's not super clear here that once you run the above docker command, you'll be dropped into a shell inside that instance and then this Rscript command you would be running there

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
## Installation

To use the software in this and related repositories (more on this below) you
need to work on developer mode. There is no "installation package"; instead
 
@cjyetman cjyetman 9 hours ago Member
nit: I don't like this "developer mode" phrase... afaik that's an undefined phrase which then leaves the reader to make up their own opinion about what that means, which could vary greatly

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
To use the software in this and related repositories (more on this below) you
need to work on developer mode. There is no "installation package"; instead
you need to clone the source code [2DII's organization on
GitHub](https://github.com/2DegreesInvesting/).
 
@cjyetman cjyetman 9 hours ago Member
should this point to https://github.com/2DegreesInvesting/PACTA_analysis?

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
```{r}
devtools::load_all()
```
Comment on lines +153 to +155
 
@cjyetman cjyetman 9 hours ago Member
I think this could use more explanation, e.g. the command devtools::load_all() loads all scripts that live in the R directory, simulating loading the code in the repo as if it were an installed package, comparable to library(PACTA_analysis)

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 9 hours ago
README.Rmd
```{bash comment="$"}
ls web_tool_script*
```
Comment on lines +170 to +172
 
@cjyetman cjyetman 9 hours ago Member
for me, running a bash command just to show 3 filenames distracts from the intent and I find myself scratching my head wondering what I'm supposed to take away from this and/or feeling like maybe I didn't fully understand what was intended here

 
@cjyetman cjyetman 8 hours ago Member
and this comes out in the HTML output as well...

ls web_tool_script*
$ web_tool_script_1.R
$ web_tool_script_2.R
$ web_tool_script_3.R
@maurolepore	Reply…
@cjyetman
cjyetman reviewed 8 hours ago
README.Rmd
you can run scripts 1 and 2 with:

```{r}
source_web_tool_scripts(1:2)
 
@cjyetman cjyetman 8 hours ago Member
I think it needs to be described here that source_web_tool_scripts is a custom function that is defined... somewhere? And that it will not be available until after the user has run devtools::load_all() (I guess?) or the script that defined it is sourced. I personally don't know off the top of my head where source_web_tool_scripts comes from.

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 8 hours ago
README.Rmd
With the terminal, you can run each script individually. For example, you can
run script 3 with:

```{bash comment="$", error=TRUE}
Rscript --vanilla web_tool_script_3.R TestPortfolio_Input  
```
Comment on lines +182 to +187
 
@cjyetman cjyetman 8 hours ago Member
will this work on Windows? I guess if someone has a bash shell emulator like git shell or something? I'm really not sure. Anyway, maybe make it more clear what is meant specifically by "terminal" and where that is or is not relevant?

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 8 hours ago
README.Rmd
* Run them on your local computer to track and explore their implementation.
* Run them on a server as an integration test.

### Single-inputs workflow
 
@cjyetman cjyetman 8 hours ago Member
There's a pretty important absence here... any explanation of where the critically important input files are or need to be. Hypothetically, the repo should work because it includes some default example files, but any analyst truly using this repo... the first thing they would want to do is put their input files in working_dir

@maurolepore	Reply…
@cjyetman
cjyetman reviewed 8 hours ago
 
Member
cjyetman left a comment
I fully reviewed README.Rmd as it currently is. It's obviously not complete so I'll submit this without explicit approval.

Merge state
Add more commits by pushing to the generalize-workflow branch on 2DegreesInvesting/PACTA_analysis.

This branch has not been deployed
No deployments
Review required
At least 1 approving review is required by reviewers with write access. 
All checks have passed
5 successful checks
@github-actions
R-CMD-check / windows-latest (release) (pull_request) Successful in 3m
Details
@github-actions
Source web-tool scripts / ubuntu-18.04 (release) (pull_request) Successful in 9m
Details
@github-actions
R-CMD-check / macOS-latest (release) (pull_request) Successful in 1m
Details
@github-actions
R-CMD-check / ubuntu-16.04 (release) (pull_request) Successful in 2m
Details
@github-actions
R-CMD-check / ubuntu-16.04 (oldrel) (pull_request) Successful in 2m
Details
Merging is blocked
Merging can be performed automatically with 1 approving review.
 or view .
@maurolepore

 
 
Leave a comment
No file chosen
Attach files by dragging & dropping, selecting or pasting them.
Remember, contributions to this repository should follow its contributing guidelines and code of conduct.
 ProTip! Add comments to specific lines under Files changed.
Reviewers
@cjyetman cjyetman
At least 1 approving review is required to merge this pull request.

Still in progress? 
AssigneesNo one—
Labels
None yet
ProjectsNone yet
MilestoneNo milestone
Linked issues
Successfully merging this pull request may close these issues.

None yet

Notifications
Customize
You’re receiving notifications because you authored the thread.
2 participants
@maurolepore
@cjyetman